subprojects/base-services/src/main/java/org/gradle/internal/time/MonotonicClock.java

```diff
@@ -30,58 +31,50 @@
  * While System.nanoTime() is usually monotonic it is not actually guaranteed, especially on virtualized hardware.
  *
  * - http://bugs.java.com/bugdatabase/view_bug.do?bug_id=6458294
- *
- * @see Time#clock()
  */
-public class MonotonicClock implements Clock {
+class MonotonicClock implements Clock {
+
+    private static final long SYNC_INTERVAL_MILLIS = TimeUnit.SECONDS.toMillis(5);
 
     private final TimeSource timeSource;
-    private final long startMillis;
-    private final long startNanos;
 
-    private final AtomicLong max = new AtomicLong();
+    private long syncMillis;
+    private long syncNanos;
+
+    private long max;
+
+    private final Lock lock = new ReentrantLock();
+    private long syncIntervalMillis;
 
-    public MonotonicClock() {
-        this(new TimeSource.True());
+    MonotonicClock() {
+        this(TimeSource.SYSTEM, SYNC_INTERVAL_MILLIS);
     }
 
     @VisibleForTesting
-    MonotonicClock(TimeSource timeSource) {
+    MonotonicClock(TimeSource timeSource, long syncIntervalMillis) {
         this.timeSource = timeSource;
-        this.startMillis = timeSource.currentTimeMillis();
-        this.startNanos = timeSource.nanoTime();
+        this.syncMillis = this.max = timeSource.currentTimeMillis();
+        this.syncNanos = timeSource.nanoTime();
+        this.syncIntervalMillis = syncIntervalMillis;
     }
 
-    @Override
     public long getCurrentTime() {
-        long elapsedMills = TimeUnit.NANOSECONDS.toMillis(timeSource.nanoTime() - startNanos);
-        long currentTime = startMillis + elapsedMills;
-        long currentMax;
-        do {
-            currentMax = max.get();
-            currentTime = Math.max(currentTime, currentMax);
-        } while (!max.compareAndSet(currentMax, currentTime));
-
-        return currentTime;
-    }
-
-    interface TimeSource {
-
-        long currentTimeMillis();
-
-        long nanoTime();
-
-        class True implements TimeSource {
-            @Override
-            public long currentTimeMillis() {
-                return System.currentTimeMillis();
-            }
-
-            @Override
-            public long nanoTime() {
-                return System.nanoTime();
+        lock.lock();
```

Comment:
Not sure about the performance impact of this, let's see what the tests say.

Reply:
Indeed. 

It can be written with a finer grained lock, or lock free (I think), but that's more than I can deal with at this time of day.

Reply:
Perf tests pass.